internal/app/upstream/client.go

```diff
@@ -55,11 +73,184 @@ type Response struct {
 //
 // The method does not block until the underlying connection is up.
 // Returns immediately and connecting the server happens in background
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	// TODO: configure grpc options.https://github.com/envoyproxy/xds-relay/issues/55
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+func (m *client) OpenStream(request *v2.DiscoveryRequest) (<-chan *v2.DiscoveryResponse, chan bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case ListenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case ClusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case RouteTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case EndpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		cancel()
+		return nil, nil, &UnsupportedResourceError{TypeURL: request.GetTypeUrl()}
+	}
+
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	signal := make(chan *version, 1)
+	// The xds protocol https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack
+	// specifies that the first request be empty nonce and empty version.
+	// The origin server will respond with the latest version.
+	signal <- &version{nonce: "", version: ""}
+
+	done := make(chan bool, 1)
+	response := make(chan *v2.DiscoveryResponse, 1)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go send(ctx, cancel, done, &wg, request, response, stream, signal, m.callOptions)
+	go recv(cancel, done, &wg, response, stream, signal)
+
+	go func() {
+		wg.Wait()
+		closeChannels(signal, response)
+	}()
+
+	return response, done, nil
+}
+
+func send(
+	ctx context.Context,
+	cancelFunc context.CancelFunc,
+	done <-chan bool,
+	wg *sync.WaitGroup,
+	request *v2.DiscoveryRequest,
+	response chan *v2.DiscoveryResponse,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(ctx, func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
+			if err != nil {
+				select {
+				case <-done:
+					cancelFunc()
```

Comment:
There is a nuance here which is coming up because of using closed channel for error signaling: Closing channel on error vs closing channel cleanly.
The `done` channel indicates there was a signal from orchestrator to cleanup. It's not an error condition anymore. The orchestrator has to behave in 2 ways, if its signaling a done, it should not wait on a closed Response channel, because it will be anyway cleanly closed. If there was no signal via done to cleanup and the Response channel was still closed, it indicates an error. 
I was of the opinion that there should be a different signaling mechanism just to indicate error, lmk what you feel about it.

Reply:
I might have misunderstood the question. I changed the implementation some more. LMK if that is aligned with what you are thinking.